File: app/src/main/java/im/point/dotty/repository/AllRepo.java
package im.point.dotty.repository;

import java.util.List;

import im.point.dotty.db.AllPostDao;
import im.point.dotty.mapper.Mapper;
import im.point.dotty.model.AllPost;
import im.point.dotty.network.MetaPost;
import im.point.dotty.network.ObservableCallBackAdapter;
import im.point.dotty.network.PointAPI;
import im.point.dotty.network.PostsReply;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

class AllRepo implements Repository<AllPost> {
    private final PointAPI api;
    private final AllPostDao allPostDao;
    private final String token;
    private final Mapper<AllPost, MetaPost> mapper;

    AllRepo(PointAPI api, String token, AllPostDao allPostDao, Mapper<AllPost, MetaPost> mapper) {
        this.api = api;
        this.allPostDao = allPostDao;
        this.token = token;
        this.mapper = mapper;
    }

    @Override
    public Flowable<List<AllPost>> getAll() {
        return allPostDao.getAll();
    }

    @Override
    public Single<List<AllPost>> fetch() {
        Observable<PostsReply> source = Observable.create(emitter -> {
            api.getAll(token, "").enqueue(new ObservableCallBackAdapter<>(emitter));
        });
        return source.observeOn(Schedulers.io())
                .flatMap(reply -> Observable.fromIterable(reply.getPosts()))
                .map(mapper::map)
                .toList()
                .doOnSuccess(allPosts -> allPostDao.insertAll(allPosts));
    }
}

File: app/src/main/java/im/point/dotty/repository/AllRepo.kt
package im.point.dotty.repository

import android.annotation.SuppressLint
import im.point.dotty.db.AllPostDao
import im.point.dotty.domain.AppState
import im.point.dotty.mapper.AllPostMapper
import im.point.dotty.mapper.Mapper
import im.point.dotty.model.AllPost
import im.point.dotty.network.MetaPost
import im.point.dotty.network.ObservableCallBackAdapter
import im.point.dotty.network.PointAPI
import im.point.dotty.network.PostsReply
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.ObservableEmitter
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

class AllRepo(private val api: PointAPI,
              private val state: AppState,
              private val allPostDao: AllPostDao,
              private val mapper: Mapper<AllPost, MetaPost> = AllPostMapper())
    : Repository<AllPost, String> {

    @SuppressLint("CheckResult")
    fun fetch(isBefore: Boolean): Single<List<AllPost>> {
        val source = Observable.create { emitter: ObservableEmitter<PostsReply> ->
            api.getAll(state.token
                    ?: throw Exception("invalid token"), if (isBefore) state.allPageId else null)
                    .enqueue(ObservableCallBackAdapter(emitter))
        }
                .observeOn(Schedulers.io())
                .flatMap { reply: PostsReply -> Observable.fromIterable(reply.posts) }
                .map { entry: MetaPost -> mapper.map(entry) }
        source.lastElement().subscribe { post -> state.allPageId = post.pageId }
        return source.toList()
                .doOnSuccess { allPosts: List<AllPost> -> allPostDao.insertAll(allPosts) }
    }

    override fun getAll(): Flowable<List<AllPost>> {
        return allPostDao.getAll()
    }

    override fun getItem(id: String): Flowable<AllPost> {
        return allPostDao.getPost(id)
    }

    override fun fetch(): Single<List<AllPost>> {
        return fetch(false)
    }

    fun fetchBefore(): Single<List<AllPost>> {
        return fetch(true)
    }
}